### ISSUE TITLE
Single Sign Out is not working, because of cross origin cookies problems.

### ISSUE THREAD

**Issue by DerAlbertCom** [IdentityServer/IdentityServer4#1288]
TL:DR: The Signout Page of the Client does not receive the authentication cookie when IdServer notifies the clients when the user signs out.

I have several Clients which use IdServer for SSO. Login works, Cookie Authentication in the Client works, also the Clients get notified when the user logs out. But, because of Browser Restrictions the Client did not receive the authentication Cookie, because of the iframe in which the LogoutUrl is embedded.

The Logout Url gets the sid and iss, and I of course i want to verify that with the current logged in user, without verifying the a logout DOS is an easy attack. But without the Cookie there is no current user.

For IE I can add the IdServer Url to the trusted Sites, and then the Cookie get's send. And the user will be logged out.

For Firefox and Chrome I can allow 3rd party Cookies, and also here the User will be correctly validated and logged out.

But I don't have control over the Browser of the User ;)

I tried some Ideas with CSP and X-Frame-Options on the Client Side, but that does not worked. Or i'm doing it wrong ;)

Any Ideas how to fix this?

**Comment by brockallen** [comment]
> Any Ideas how to fix this?

Same-origin cookies was not part of the design in OIDC. Don't use them, and use more traditional anti-xsrf measures.

**Comment by DerAlbertCom** [comment]
Well,  I'm not sure that i understand you. I'm don't use any special cookies, simple the Auth Cookie (from ASP.NET Katana). Katana has no OpenId Connect SignOut  from the Identity Provider.

IdServer seems it does a frontchannel logout for the RP. On my side i try to do this.

http://openid.net/specs/openid-connect-frontchannel-1_0.html#OPLogout

```c#
        private Task GlobalLogoutHandler(IOwinContext context)
        {
            context.Response.Headers["Cache-Control"] = "no-cache, no-store";
            context.Response.Headers["Pragma"] = "no-cache";
            var user = context.Request.User as ClaimsPrincipal;
            if (user == null)
            {
                return CompletedTask;
            }
            var sid = context.Request.Query.Get("sid");
            var iss = context.Request.Query.Get("iss");
            var sidClaim = user.FindFirst("sid");
            var issClaimd = user.FindFirst("iss");
            if (sidClaim?.Value == sid && issClaimd?.Value == iss)
            {
                context.Authentication.SignOut(CookieAuthenticationDefaults.AuthenticationType);
            }
            else
            {
                context.Response.StatusCode = 420;
            }
            return CompletedTask;
        }
```

Simply validating iss and sid against the current autneticated User (this values are in the cookie). But there is no authenticated use because of the not sent cookie.

I don't have an Idea.

**Comment by brockallen** [comment]
Ok, sorry -- I think I now understand. So because the end user has done this setting in the browser the cookies is not sent in the iframe. So yes, you're right -- this is a possible problem that the spec does not address.

**Comment by DerAlbertCom** [comment]
So, and how do I now a Single Signout?

**Comment by brockallen** [comment]
Is single sign-out not working?

**Comment by DerAlbertCom** [comment]
No, not with Katana ... the Logout Uri will get called, but i can't identify the user to verify the signout.

**Comment by brockallen** [comment]
Are you passing the id_token_hint as in our examples?

**Comment by DerAlbertCom** [comment]
Sorry, maybe I'm not clear enough. There is NO problem with IdentityServer 4 with that.
I give the Token hint, the User will be logged out from IdServer4 an the LogoutUri will be put in the IFrame on your side and MY site must now log out.

But OpenId Connect with Katana has not implemented signout over for this. This was done with ASP.NET Core OpenId Connect. Core has a /signout-oidc, Katana has not suche Uri.

But i need to logout, so I asked the Question here, because i thought you have plenty of knowledge in that area.

So, what can I do? Any Ideas.

**Comment by brockallen** [comment]
We have an example in out IdSvr3 client samples.

**Comment by DerAlbertCom** [comment]
Yes .... https://github.com/IdentityServer/IdentityServer3.Samples/blob/master/source/Clients/MVC%20OWIN%20Client%20(Hybrid)/Controllers/HomeController.cs but with this, we are on my starting Question ... no Cookies, no Auth, no User, not possible to Identity the validity of the Signout.

**Comment by leastprivilege** [comment]
Right. You are out of luck then.

For v2 we plan to add the back-channel logout protocol. This would be the solution to the problem. But for front-channel is all you have and it always has been a best-effort approach.
